### PR TITLE
fix(connection): fixing mssql reconnect bug

### PR DESCRIPTION
**Pull request by codyseibert**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in CONTRIBUTING.md?

### Description of change
Original Issue:
https://github.com/sequelize/sequelize/issues/8014

There is currently a bug where if you restart a mssql database,
sequelize will lock the pool resources.  Once those are locked, all
future requests to .acquire will throw a ResourceRequest timed out
error.  This commit wraps all the connect logic in a retry to prevent an
inifite loop of trying to connect, and also watchins for an 'end' event
in the mssql adapter to properly reject the promise if 'end' is sent
before 'connect' is fired.

### CODE REVIEW

**Comment by codyseibert** [comment]
Anyone want to enlighten me as to what this "codecov/patch" failure is?  I assume I need to have 95% coverage over the changes I added or something?

File: lib/dialects/abstract/connection-manager.js [hunk 1]
```diff
@@ -84,18 +86,13 @@ class ConnectionManager {
...
-          this
-            ._connect(config)
-            .tap(() => {
-              this.poolError = null;
-            })
+        create: () => new Promise((resolve, reject) => {
+          retry(() => this._connect(config), {
+            max: config.pool.connectRetries
+          })
             .then(resolve)
```
Comment (eseliger) [inline lib/dialects/abstract/connection-manager.js]: you can just return the promise, no need to wrap into another

File: lib/dialects/abstract/connection-manager.js [hunk 2]
```diff
@@ -173,16 +170,16 @@ class ConnectionManager {
...
+              max: config.pool.connectRetries
+            })
+              .then(connection => {
                 connection.queryType = 'read';
-                this.poolError = null;
                 resolve(connection);
               })
               .catch(e => {
-                this.poolError = e;
+                reject(e);
```
Comment (eseliger) [inline lib/dialects/abstract/connection-manager.js]: same here

File: lib/dialects/postgres/connection-manager.js [hunk 3]
```diff
@@ -110,17 +110,20 @@ class ConnectionManager extends AbstractConnectionManager {
...
             reject(new sequelizeErrors.ConnectionError(err));
           }
+          connection.end();
           return;
         }
         responded = true;
         debug('connection acquired');
         resolve(connection);
       });
+
```
Comment (eseliger) [inline lib/dialects/postgres/connection-manager.js]: unnecessary new line

**Comment by eseliger** [comment]
@codyseibert What happened here? 🙈

**Comment by codyseibert** [comment]
@eseliger I don't know how to re-run the tests without closing and opening the PR.  All these tests passed once before, but now they seem to be failing on arbitrary failures.

**Comment by eseliger** [comment]
You can amend on your current commit and then force push on your branch to retrigger this

**Comment by codyseibert** [comment]
ok, I will try that for now on.  Any clue what these coverage failures are?

**Comment by eseliger** [comment]
You mean the codecov check? 
It's purpose is to remind you the average coverage is 95.34%, your PR just has 81.25% coverage on changed or added lines.

**Comment by codyseibert** [comment]
that, but more specifically the "continuous-integration/appveyor/pr" and "continuous-integration/travis-ci/pr" jobs fail randomly at times, but pass on other runs:

appveyor
```
npm ERR! Windows_NT 6.3.9600
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "test-integration"
npm ERR! node v4.8.4
npm ERR! npm  v2.15.11
npm ERR! code ELIFECYCLE
npm ERR! sequelize@0.0.0-development test-integration: `mocha --globals setImmediate,clearImmediate --ui tdd --check-leaks --colors -t 30000 --reporter spec "test/integration/**/*.test.js"`
npm ERR! Exit status 3221225477
npm ERR! 
npm ERR! Failed at the sequelize@0.0.0-development test-integration script 'mocha --globals setImmediate,clearImmediate --ui tdd --check-leaks --colors -t 30000 --reporter spec "test/integration/**/*.test.js"'.
npm ERR! This is most likely a problem with the sequelize package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     mocha --globals setImmediate,clearImmediate --ui tdd --check-leaks --colors -t 30000 --reporter spec "test/integration/**/*.test.js"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs sequelize
npm ERR! Or if that isn't available, you can get their info via:
npm ERR! 
npm ERR!     npm owner ls sequelize
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     C:\projects\sequelize\npm-debug.log
npm ERR! Test failed.  See above for more details.
Command exited with code 1
```

**Comment by eseliger** [comment]
Unfortunately, some times the tests fail due to timeouts and so on, for now, restarting them is the simplest solution, which is most easily done by just triggering another push as I wrote 2 messages ago

**Comment by codyseibert** [comment]
@eseliger what do you recommend I run in npm to test and increase my codecov/patch coverage?  I added a unit tests but it didn't seem to increase my codecov/patch report...

**Comment by codyseibert** [comment]
@eseliger I'm not sure why it says line 77 is no covered.  I wrote a unit tests to cover that specifically and it passes fine.

**Comment by Vratislav** [comment]
@codyseibert we have tested this in our codebase and it solves our issue we are facing when the sequelize never reconnects to the database that was restarted.

We have also tested both https://github.com/sequelize/sequelize/pull/7924 and https://github.com/sequelize/sequelize/pull/7922 None of those solve the `ResourceRequest timed out` that we are facing. 

@eseliger @codyseibert Can we help in any way to get this merged?

**Comment by prokopsimek** [comment]
Still broken. The same issue `ResourceRequest timed out`.

**Comment by prokopsimek** [comment]
Possible ref.: https://github.com/coopernurse/node-pool/issues/155

**Comment by smithdalec** [comment]
This fix solves the `ResourceRequest timed out` error for me. Is there anything in particular we're waiting on in order to merge?

**Comment by depoulo** [comment]
This fix solves the `ResourceRequest timed out` error for me as well (MySQL/MariaDB here). Switched to using the PR in my `package.json` for now, hope this will be merged some day. After all it's a severe regression from v3.x

**Comment by jasonfma** [comment]
We are still seeing this error as well. Any update here?

**Comment by tylerjwatson** [comment]
We've had a couple of problems that appear to be around this in production, the latest one just now where to recover the pool we've had to restart the server.  It looks like the pool has almost completely drained apart from a couple of active connections.

I know it seems like nagging at this point, but it's becoming a pressing issue.  Can we please kick the can along on this patch again?  Is there anything I can do to help it get across the line?

**Comment by benjohnson** [comment]
We're also in the same boat. I'm not sure what else is needed to get this merged, but we're happy to lend a hand if there's something in particular this PR is missing.

**Comment by ADumaine** [comment]
Is there a workaround for this while we are waiting for it to be merged?

This is affecting our tracking system.  It is growing fast and the dead connections are becoming more of an issue every day.

**Comment by aviranco** [comment]
This fix is also relevant for PostgreSQL or only for mysql?

**Comment by depoulo** [comment]
@ADumaine [we're still at this state](https://github.com/sequelize/sequelize/pull/8022#issuecomment-326011895), yes, since August!

> Switched to using the PR in my package.json for now

**Comment by guyellis** [comment]
Has this PR been superseded by https://github.com/sequelize/sequelize/pull/8815?

**Comment by depoulo** [comment]
@guyellis looks like. Currently updating my project from using this PR to using the new one (and adopting all changes that happened to Sequelize in the past 6 months). The new PR fixes the issue for me.

**Pull request closed without merging**